clippy_lints/src/undocumented_unsafe_block_safety.rs

```diff
@@ -0,0 +1,198 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::{snippet, indent_of, reindent_multiline};
+use rustc_errors::Applicability;
+use rustc_hir::{Block, BlockCheckMode, ExprKind, HirId, Local, UnsafeSource};
+use rustc_lexer::TokenKind;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty::TyCtxt;
+use rustc_session::{impl_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+use core::str::Lines;
+use core::iter::Rev;
+use std::borrow::Cow;
+use clippy_utils::is_lint_allowed;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `unsafe` blocks without a `// Safety: ` comment
+    /// explaining why the unsafe operations performed inside
+    /// the block are (or are not) safe.
+    ///
+    /// ### Why is this bad?
+    /// Unsafe blocks without a safety comment can be difficult to
+    /// understand for those that are unfamiliar with the block's code
+    /// and/or unsafe in general.
+    ///
+    /// ### Example
+    /// ```rust
+    ///   unsafe {};
+    ///
+    ///   let _ = unsafe {};
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // Safety: ...
+    /// unsafe {}
+    ///
+    /// // Safety: ...
+    /// let _ = unsafe {};
+    /// ```
+    pub UNDOCUMENTED_UNSAFE_BLOCK_SAFETY,
+    restriction,
+    "creating an unsafe block without explaining why it is or is not safe"
+}
+
+impl_lint_pass!(UndocumentedUnsafeBlockSafety => [UNDOCUMENTED_UNSAFE_BLOCK_SAFETY]);
+
+#[derive(Default)]
+pub struct UndocumentedUnsafeBlockSafety {
+    pub local: bool
+}
+
+#[derive(PartialEq)]
+enum CommentType {
+    Safety,
+    Unrelated,
+    None,
+}
+
+impl LateLintPass<'_> for UndocumentedUnsafeBlockSafety {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &'_ Block<'_>) {
+        if self.local {
+            self.local = false;
+            return
+        }
+
+        if_chain! {
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCK_SAFETY, block.hir_id);
+            if !in_external_macro(cx.tcx.sess, block.span);
+            if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules;
+            if let Some(enclosing_scope_hir_id) = cx.tcx.hir().get_enclosing_scope(block.hir_id);
+            then {
+                find_candidate(cx, block.span, enclosing_scope_hir_id)
+            }
+        }
+    }
+
+    fn check_local(&mut self, cx: &LateContext<'_>, local: &'_ Local<'_>) {
+        if_chain! {
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCK_SAFETY, local.hir_id);
+            if !in_external_macro(cx.tcx.sess, local.span);
+            if let Some(init) = local.init;
+            if let ExprKind::Block(block, _) = init.kind;
```

Comment:
This won't work for 

```rs
let _ = (42, unsafe { *ptr::null_ptr() });
```

Please also add a test for this.

Reply:
Do you mean a UI test, or should it be handled?

Reply:
Putting a `SAFETY` comment above this line should also be acceptable.